src/C-interface/dense/bml_normalize_dense_typed.c

```diff
@@ -69,40 +69,68 @@ void *TYPED_FUNC(
 
     double *eval = bml_allocate_memory(sizeof(double) * 2);
 
-#pragma omp parallel for \
+#ifdef _OPENMP
+    int num_threads = omp_get_num_threads();
+#else
+    int num_threads = 1;
+#endif
+
+    double _emin[num_threads];
+    double _emax[num_threads];
+
+#pragma omp parallel for shared(_emin, _emax)
+    for (int i = 0; i < num_threads; i++)
+    {
+        _emin[i] = emin;
+        _emax[i] = emax;
+    }
+
+#pragma omp parallel \
     default(none) \
     shared(N, A_matrix) \
     shared(A_localRowMin, A_localRowMax, myRank) \
-    private(absham, radius, dvalue) \
-    reduction(max:emax) \
-    reduction(min:emin)
-    //for (int i = 0; i < N; i++)
-    for (int i = A_localRowMin[myRank]; i < A_localRowMax[myRank]; i++)
+    shared(_emin, _emax) \
+    private(absham, radius, dvalue)
     {
-        radius = 0.0;
+#ifdef _OPENMP
+        int thread_num = omp_get_thread_num();
+#else
+        int thread_num = 1;
+#endif
+
+#pragma omp for
```

Comment:
The `for` construct here implies a barrier after the loop. We can be certain that the serial section further down that collects the thread local min/max values is executed only after all threads are done.